File: app/src/main/java/com/metehanyl/calarsaat/data/PrefsManager.kt
package com.metehanyl.calarsaat.data

import android.content.Context

class PrefsManager(context: Context) {

    private val prefs = context.getSharedPreferences("alarm_prefs", Context.MODE_PRIVATE)

    fun isPinSet(): Boolean = prefs.contains(KEY_PIN_HASH)

    fun setPin(pin: String) {
        prefs.edit().putString(KEY_PIN_HASH, PinHasher.hash(pin)).apply()
    }

    fun verifyPin(pin: String): Boolean {
        val stored = prefs.getString(KEY_PIN_HASH, null) ?: return false
        return stored == PinHasher.hash(pin)
    }

    fun isSabahNamaziEnabled(): Boolean = prefs.getBoolean(KEY_SABAH_NAMAZI_ENABLED, false)

    fun setSabahNamaziEnabled(enabled: Boolean) {
        prefs.edit().putBoolean(KEY_SABAH_NAMAZI_ENABLED, enabled).apply()
    }

    fun getSabahNamaziAlarmCount(): Int = prefs.getInt(KEY_SABAH_COUNT, DEFAULT_SABAH_COUNT)

    fun setSabahNamaziAlarmCount(count: Int) {
        prefs.edit().putInt(KEY_SABAH_COUNT, count).apply()
    }

    fun getSabahNamaziIntervalMinutes(): Int = prefs.getInt(KEY_SABAH_INTERVAL, DEFAULT_SABAH_INTERVAL)

    fun setSabahNamaziIntervalMinutes(minutes: Int) {
        prefs.edit().putInt(KEY_SABAH_INTERVAL, minutes).apply()
    }

    fun getSabahNamaziOffsetMinutes(): Int = prefs.getInt(KEY_SABAH_OFFSET, DEFAULT_SABAH_OFFSET)

    fun setSabahNamaziOffsetMinutes(minutes: Int) {
        prefs.edit().putInt(KEY_SABAH_OFFSET, minutes).apply()
    }

    fun isSabahNamaziPinRequired(): Boolean = prefs.getBoolean(KEY_SABAH_REQUIRE_PIN, false)

    fun setSabahNamaziPinRequired(required: Boolean) {
        prefs.edit().putBoolean(KEY_SABAH_REQUIRE_PIN, required).apply()
    }

    fun getSabahNamaziPinHash(): String? = prefs.getString(KEY_SABAH_PIN_HASH, null)

    fun setSabahNamaziPinHash(hash: String?) {
        prefs.edit().putString(KEY_SABAH_PIN_HASH, hash).apply()
    }

    fun getSabahNamaziVolume(): Int = prefs.getInt(KEY_SABAH_VOLUME, DEFAULT_SABAH_VOLUME)

    fun setSabahNamaziVolume(volume: Int) {
        prefs.edit().putInt(KEY_SABAH_VOLUME, volume).apply()
    }

    fun isOemAutostartPromptShown(): Boolean = prefs.getBoolean(KEY_OEM_AUTOSTART_PROMPT_SHOWN, false)

    fun setOemAutostartPromptShown(shown: Boolean) {
        prefs.edit().putBoolean(KEY_OEM_AUTOSTART_PROMPT_SHOWN, shown).apply()
    }

    companion object {
        private const val KEY_PIN_HASH = "pin_hash"
        private const val KEY_SABAH_NAMAZI_ENABLED = "sabah_namazi_enabled"
        private const val KEY_SABAH_COUNT = "sabah_namazi_count"
        private const val KEY_SABAH_INTERVAL = "sabah_namazi_interval"
        private const val KEY_SABAH_OFFSET = "sabah_namazi_offset"
        private const val KEY_SABAH_REQUIRE_PIN = "sabah_namazi_require_pin"
        private const val KEY_SABAH_PIN_HASH = "sabah_namazi_pin_hash"
        private const val KEY_SABAH_VOLUME = "sabah_namazi_volume"
        private const val KEY_OEM_AUTOSTART_PROMPT_SHOWN = "oem_autostart_prompt_shown"

        // Default settings reproduce the original fixed behavior: imsak+10, +14, +18.
        const val DEFAULT_SABAH_COUNT = 3
        const val DEFAULT_SABAH_INTERVAL = 4
        const val DEFAULT_SABAH_OFFSET = 10
        const val DEFAULT_SABAH_VOLUME = 100
    }
}

File: ios/CalarSaat/Persistence/PrefsManager.swift
import Foundation

/// Mirrors Android's `PrefsManager` (SharedPreferences `alarm_prefs`) using `UserDefaults`.
final class PrefsManager {
    static let shared = PrefsManager()

    private let defaults = UserDefaults.standard

    private enum Key {
        static let pinHash = "pin_hash"
        static let sabahNamaziEnabled = "sabah_namazi_enabled"
        static let sabahNamaziCount = "sabah_namazi_count"
        static let sabahNamaziInterval = "sabah_namazi_interval"
        static let sabahNamaziOffset = "sabah_namazi_offset"
        static let sabahNamaziRequirePin = "sabah_namazi_require_pin"
        static let sabahNamaziPinHash = "sabah_namazi_pin_hash"
    }

    // MARK: Global PIN

    var isPinSet: Bool { defaults.string(forKey: Key.pinHash) != nil }

    func setPin(_ pin: String) {
        defaults.set(PinHasher.hash(pin), forKey: Key.pinHash)
    }

    func verifyPin(_ pin: String) -> Bool {
        PinHasher.verify(pin, against: defaults.string(forKey: Key.pinHash))
    }

    // MARK: Sabah Namazı settings

    var isSabahNamaziEnabled: Bool {
        get { defaults.bool(forKey: Key.sabahNamaziEnabled) }
        set { defaults.set(newValue, forKey: Key.sabahNamaziEnabled) }
    }

    var sabahNamaziAlarmCount: Int {
        get { defaults.object(forKey: Key.sabahNamaziCount) as? Int ?? 3 }
        set { defaults.set(newValue, forKey: Key.sabahNamaziCount) }
    }

    var sabahNamaziIntervalMinutes: Int {
        get { defaults.object(forKey: Key.sabahNamaziInterval) as? Int ?? 4 }
        set { defaults.set(newValue, forKey: Key.sabahNamaziInterval) }
    }

    var sabahNamaziOffsetMinutes: Int {
        get { defaults.object(forKey: Key.sabahNamaziOffset) as? Int ?? 10 }
        set { defaults.set(newValue, forKey: Key.sabahNamaziOffset) }
    }

    var isSabahNamaziPinRequired: Bool {
        get { defaults.bool(forKey: Key.sabahNamaziRequirePin) }
        set { defaults.set(newValue, forKey: Key.sabahNamaziRequirePin) }
    }

    var sabahNamaziPinHash: String? {
        get { defaults.string(forKey: Key.sabahNamaziPinHash) }
        set { defaults.set(newValue, forKey: Key.sabahNamaziPinHash) }
    }

    func setSabahNamaziPin(_ pin: String) {
        sabahNamaziPinHash = PinHasher.hash(pin)
    }
}
